fn/goroutine_manager_test.go

```diff
@@ -0,0 +1,109 @@
+package fn
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+// TestGoroutineManager tests that GoroutineManager starts goroutines, until ctx
+// expires fails to start after it expires and waits for already started
+// goroutines in Stop method.
+func TestGoroutineManager(t *testing.T) {
+	t.Parallel()
+
+	m := NewGoroutineManager(context.Background())
+
+	taskChan := make(chan struct{})
+
+	require.NoError(t, m.Go(func(ctx context.Context) {
+		<-taskChan
+	}))
+
+	t1 := time.Now()
+
+	// Close taskChan in 1s, causing the goroutine to stop.
+	time.AfterFunc(time.Second, func() {
+		close(taskChan)
+	})
+
+	m.Stop()
+	stopDelay := time.Since(t1)
+
+	// Make sure Stop was waiting for the goroutine to stop.
+	require.Greater(t, stopDelay, time.Second)
+
+	// Make sure new goroutines do not start after Stop.
+	require.ErrorIs(t, m.Go(func(ctx context.Context) {}), ErrStopping)
+}
+
+// TestGoroutineManagerContextExpires tests the effect of context expiry.
+func TestGoroutineManagerContextExpires(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	m := NewGoroutineManager(ctx)
+
+	require.NoError(t, m.Go(func(ctx context.Context) {
+		<-ctx.Done()
+	}))
+
+	// The context of the manager should not expire, so the following call
+	// must block.
+	select {
+	case <-m.Context().Done():
+		t.Errorf("context must not expire at this point")
+	default:
+	}
+
+	cancel()
+
+	// The context of the manager should expire, so the following call
+	// must not block.
+	select {
+	case <-m.Context().Done():
+	default:
+		t.Errorf("context must expire at this point")
+	}
+
+	// Make sure new goroutines do not start after context expiry.
+	require.ErrorIs(t, m.Go(func(ctx context.Context) {}), ErrStopping)
+
+	// Stop will wait for all goroutines to stop.
+	m.Stop()
+}
+
+// TestGoroutineManagerStress starts many goroutines while calling Stop.
+func TestGoroutineManagerStress(t *testing.T) {
```

Comment:
It appears there are no conditions being checked in this test.

Reply:
probably we can check for the recovered panic ?

Reply:
This test makes sure that there is no race condition in WaitGroup usage. If the mutex is removed in the implementation, this test crashes with some probability under `-race` (I run it with `-count 100` to get some failures).

@ziggie1984 If panic happen, the test fails already.

Reply:
I added to the test godoc:

```go
// TestGoroutineManagerStress starts many goroutines while calling Stop. It
// is needed to make sure the GoroutineManager does not crash if this happen.
// If the mutex was not used, it would crash because of a race condition between
// wg.Add(1) and wg.Wait().
func TestGoroutineManagerStress(t *testing.T) {
```